pkgs/applications/misc/web-media-controller/default.nix

```diff
@@ -0,0 +1,34 @@
+{ stdenv, fetchFromGitHub, cmake, pkgconfig, glib, pcre, json-glib }:
+
+stdenv.mkDerivation rec {
+  pname = "wmc-mpris";
+  name = "${pname}-unstable";
+  version = "2019-04-12";
+
+  src = fetchFromGitHub {
+    owner = "f1u77y";
+    repo = pname;
+    rev = "fcfb1d8f708b8e69bc2ab8b9bc0e2605d9ace585";
+    sha256 = "0plknl1r3z58sl68s60z4cxlwk7z5ld09lza74x5m8knhr94vwy9";
+  };
+
+  nativeBuildInputs = [ cmake pkgconfig ];
+  buildInputs = [ glib pcre json-glib ];
+  # Without these, the build fails because cmake tries to create files in /etc/
+  # These files are meant for the web extensions' native messaging host configuration.
+  # upstream: https://github.com/f1u77y/wmc-mpris/blob/master/me.f1u77y.web_media_controller.firefox.json
+  # Nix users will have to create these files by themselves.
+  cmakeFlags = [
+    "-DENABLE_CHROMIUM=Off"
```

Comment:
We should be able to install these, the only problem is the paths https://github.com/f1u77y/wmc-mpris/blob/master/CMakeLists.txt#L15.

How about `substituteInPlace` them in `postPatch` to be prefixed with `$out`?

Reply:
Where would we put them then?

Reply:
These locations are correct
https://github.com/f1u77y/wmc-mpris/blob/fcfb1d8f708b8e69bc2ab8b9bc0e2605d9ace585/CMakeLists.txt#L15

and `INSTALL_FOR_CURRENT_USER` will be off.
We just need to prefix those paths into out so they get installed into the nix store.

Reply:
So the final location will be `$out/etc/...`?

Reply:
Yes, and then a user in nixos could use `environment.etc`. See the chrome-gnome-shell nixos module.

Reply:
From some reason I get this tree:

```
/nix/store/8m617gad3fz0vwm1n833807dhwb03fs3-wmc-mpris-unstable
├── bin
│   └── web-media-controller
├── etc
│   ├── chromium
│   │   └── native-messaging-hosts
│   │       └── me.f1u77y.web_media_controller.json
│   └── var
│       └── empty
│           └── chrome
│               └── native-messaging-hosts
│                   └── me.f1u77y.web_media_controller.json
└── lib
    └── mozilla
        └── native-messaging-hosts
            └── me.f1u77y.web_media_controller.json

11 directories, 4 files
```

That `$out/etc/var/empty/chrome` doesn't feel right.. I've used the following `postPatch`:

```
    substituteInPlace CMakeLists.txt \
      --replace /etc/ $out/etc/ \
      --replace /usr/lib/ $out/lib/

```

Reply:
Huh, it would be nice if upstream used https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html so we didn't have to do this.

Since it's literally coping a file to a directory, how about just doing what the `CMakeLists.txt` should be doing in `postInstall`?

Reply:
[I've filed a PR upstream](https://github.com/f1u77y/wmc-mpris/pull/5) Let's give it a few days and see afterwards..

Reply:
I don't know much about NixOS but I think you could prefix the paths as I do it for [AUR](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=web-media-controller-mpris-git#n55).

Reply:
@f1u77y it's very kind of you to respond here :) If I understand you correctly, you are suggesting to use `DESTDIR` instead of `PREFIX` right? As a fairly beginner NixOS user, I can testify that to my understanding this is the essence of NixOS' functional programming - it is better explained [here](https://nixos.org/nixos/nix-pills/working-derivation.html#idm140737316223792).

If wmc-mpris would have used standard [GNU install dirs](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html), as @worldofpeace has suggested in [his previous comment](https://github.com/NixOS/nixpkgs/pull/64891#discussion_r306501728), setting `-DCMAKE_INSTALL_PREFIX` in our build according to our casual preferences would have been enough and no further work was to be done on your side.

[The PR I've added in your repo](https://github.com/f1u77y/wmc-mpris/pull/5) is a mere starting point towards a complete compatibility with GNU's standard install directories. Thanks for merging it :)

@worldofpeace I have to get going now so I will update the revision here shortly.

Reply:
> it's very kind of you to respond here :) If I understand you correctly, you are suggesting to use DESTDIR instead of PREFIX right? As a fairly beginner NixOS user, I can testify that to my understanding this is the essence of NixOS' functional programming - it is better explained here.

We could do that, but I tend to avoid this.

> If wmc-mpris would have used standard GNU install dirs, as @worldofpeace has suggested in his previous comment, setting -DCMAKE_INSTALL_PREFIX in our build according to our casual preferences would have been enough and no further work was to be done on your side.

That's true. btw it's them not he :smile:

Reply:
> That's true. btw it's them not he :smile:

So you are actually a team?

I've pushed an update that resolves to the following tree:

```
/nix/store/csf72019crdks38g2iahr4dda7rly7y7-wmc-mpris-unstable
├── bin
│   └── web-media-controller
├── etc
│   ├── chromium
│   │   └── native-messaging-hosts
│   │       └── me.f1u77y.web_media_controller.json
│   └── opt
│       └── chrome
│           └── native-messaging-hosts
│               └── me.f1u77y.web_media_controller.json
└── usr
    └── lib
        └── mozilla
            └── native-messaging-hosts
                └── me.f1u77y.web_media_controller.json

11 directories, 4 files
```

Reply:
Could you help me a bit with migrating to GNUInstallDirs? AFAICU I should use `CMAKE_INSTALL_FULL_SYSCONFDIR` instead of `/etc` and `CMAKE_INSTALL_FULL_LIBDIR` instead of `/usr/lib`. But `CMAKE_INSTALL_FULL_SYSCONFDIR` is always `/etc`, not depending on `CMAKE_INSTALL_PREFIX`, but `CMAKE_INSTALL_FULL_LIBDIR` is set accordingly (at least on Arch). What am I doing wrong?

Reply:
I think your program is kind of of an exception in that sense. What's unique, is that the browsers' installation directories of these files differ from different installation method of each. Therefor, unless you have a reliable way to detect that (e.g if Firefox for example had a Cmake module / pkg-config file that would indicate that), you could know what is the native messaging manifest directory reliably.

Since I'm not sure such a mechanism exists, perhaps it's not too bad to **almost** hard-code the `/etc/` and `/usr/lib/` paths and allow the users to use the new options I've added in the PR.